mapiproxy/libmapistore/backends/indexing_mysql.c

```diff
@@ -29,16 +30,302 @@
 #include "libmapi/libmapi_private.h"
 #include <mysql/mysql.h>
 #include "../../util/mysql.h"
+#include "../../util/ccan/hash/hash.h"
 #include "../../util/schema_migration.h"
 #include "mapiproxy/libmapiproxy/backends/openchangedb_mysql.h"
 
 #include <talloc.h>
-
+#include <libmemcached/memcached.h>
 
 #define MYSQL(context)	((MYSQL *)context->data)
 
 
 /**
+   \details Generate key for FMID cache
+
+   \param mem_ctx pointer to the memory context
+   \param uri pointer to the uri to use for hashing
+
+   \return String allocated with TALLOC on success, otherwise NULL
+ */
+static char *_memcached_gen_key(TALLOC_CTX *mem_ctx, const char *uri)
+{
+	return talloc_asprintf(mem_ctx, "%"PRIx64, hash64_any((void *)uri, strlen(uri), 0));
+}
+
+
+/**
+   \details Generate value for FMID cache
+
+   \param mem_ctx pointer to the memory context
+   \param fmid the fmid to convert
+
+   \return String allocated with TALLOC on success, otherwise NULL
+ */
+static char *_memcached_gen_value(TALLOC_CTX *mem_ctx, uint64_t fmid)
+{
+	return talloc_asprintf(mem_ctx, "%"PRIu64, fmid);
+}
+
+
+/**
+   \details Prepare FMID cache for specified user
+
+   \param ictx valid pointer to the indexing context
+   \param conn_str connection string to memcached server
+   \param username name of the user for to create the FMID cache for
+
+   \note Fallback to 127.0.0.1:11211 if server_string is missing
+
+   \return valid memcached_st pointer on success, otherwise NULL
+ */
+static memcached_st *_memcached_setup(struct indexing_context *ictx,
+				      const char *conn_str,
+				      const char *username)
+{
+	TALLOC_CTX		*mem_ctx;
+	char			*sql;
+	int			mret;
+	MYSQL_RES		*res;
+	uint32_t		num_rows = 0;
+	memcached_server_st	*servers = NULL;
+	memcached_st		*memc = NULL;
+	memcached_return	rc;
+	uint32_t		i;
+
+	DEBUG(5, ("[INFO] _memcached_setup for '%s'\n", username));
```

Comment:
Not really a fan of adding a label [INFO] (or [ERR], or [LOG LEVEL]) to the message itself. This should be done by the logging api.